kong/runloop/wasm.lua

```diff
@@ -0,0 +1,646 @@
+local _M = {}
+
+local utils = require "kong.tools.utils"
+local dns = require "kong.tools.dns"
+local clear_tab = require("table.clear")
+
+---@module 'resty.http.proxy_wasm'
+local proxy_wasm
+
+local kong = _G.kong
+local ngx = ngx
+local log = ngx.log
+local DEBUG = ngx.DEBUG
+local ERR = ngx.ERR
+local concat = table.concat
+local insert = table.insert
+local sha256 = utils.sha256_bin
+
+
+local VERSION_KEY = "filter_chains:version"
+local TTL_ZERO = { ttl = 0 }
+local ATTACH_OPTS = {}
+
+
+---
+-- Fetch the current version of the filter chain state from cache
+--
+---@return string
+local function get_version()
+  return kong.core_cache:get(VERSION_KEY, TTL_ZERO, utils.uuid)
+end
+
+
+---@alias kong.wasm.filter_chain_type
+---| 0 # service
+---| 1 # route
+---| 2 # combined
+
+local TYPE_SERVICE  = 0
+local TYPE_ROUTE    = 1
+local TYPE_COMBINED = 2
+
+
+local ENABLED = false
+
+local hash_chain
+do
+  local HASH_DISABLED = sha256("disabled")
+  local HASH_NONE     = sha256("none")
+
+  local buf = {}
+
+  ---@param chain kong.db.schema.entities.filter_chain
+  ---@return string
+  local function hash_chain_entity(chain)
+    if not chain then
+      return HASH_NONE
+
+    elseif not chain.enabled then
+      return HASH_DISABLED
+    end
+
+    local n = 0
+    for _, filter in ipairs(chain.filters) do
+      buf[n + 1] = filter.name
+      buf[n + 2] = tostring(filter.enabled)
```

Comment:
Do we still localize globals? If so: `tostring`, `ipairs`, `type`, `assert` aren't.